src/highrl/envs/env_generator.py

```diff
@@ -0,0 +1,248 @@
+"""Implementation for the environment to train the environmenet generator modle
+
+The model is assumed to take the difficulty (generated by the teacher policy) and
+generate a robot-environment abiding by the generated difficulty.
+
+In this module, we implement the environment generator using reinforcement learninig,
+for not constraining the creativity the model, i.e. there are numerous ways to generate
+environments that abide by a certain difficulty.
+"""
+from typing import Tuple, Callable, List
+import logging
+import random
+from gym import Env, spaces
+import numpy as np
+import torch
+from torch.utils.tensorboard import SummaryWriter  # type: ignore
+from torch import nn
+from stable_baselines3 import PPO  # type: ignore
+from stable_baselines3.common.policies import ActorCriticPolicy
+
+from highrl.obstacle.obstacles import Obstacles
+from highrl.obstacle.single_obstacle import SingleObstacle
+from highrl.agents.robot import Robot
+from highrl.utils import Position
+from highrl.utils.teacher_checker import compute_difficulty
+from highrl.utils.logger import init_logger
+
+_LOG = logging.getLogger(__name__)
+
+
+class EnvGeneratorPolicy(nn.Module):
+    """Implementation for the environment generator model"""
+
+    hidden_size = 16
+    hard_obs = 2
+    med_obs = 3
+    small_obs = 4
+    device = "cuda"
+    output_size = 4
+
+    def __init__(
+        self,
+        feature_dim: int = 1,
+        last_layer_dim_pi: int = 40,
+        last_layer_dim_vf: int = 32,
+    ) -> None:
+        super().__init__()
+        # IMPORTANT:
+        # Save output dimensions, used to create the distributions
+        self.latent_dim_pi = last_layer_dim_pi
+        self.latent_dim_vf = last_layer_dim_vf
+
+        self.value_net = nn.Sequential(
+            nn.Linear(feature_dim, last_layer_dim_vf),
+            nn.ReLU(),
+        )
+
+        # This linear layer will be used to map from the hidden dimension
+        # to the output dimension.
+        self.out = nn.Linear(self.hidden_size, self.output_size)
+
+        self.lstm = nn.LSTM(self.output_size, self.hidden_size)
+
+    def forward(self, features: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Forward pass through the value network and policy network"""
+        return self.forward_actor(features), self.forward_critic(features)
+
+    def _init_hidden(self, batch_size: int = 1) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Initialize the hidden and cell gate tensors for the LSTM layer
+
+        Note here we are assuming that the `num_layers` and `batch_size`
+        are equal to one.
+        """
+        return (
+            torch.zeros(1, batch_size, self.hidden_size, device=self.device),
+            torch.zeros(1, batch_size, self.hidden_size, device=self.device),
```

Comment:
What do you mean by assuming?
would not it work with another `batch_size` or `num_layers`?

Reply:
"batch_size equal to one" is a mistake here, it's not equal to one. However, for the `num_layers` it can be modified, but I don't think we need to, since the model input/output is very simple.